File: 0889-buddy-strings/0889-buddy-strings.go
func buddyStrings(s string, goal string) bool {
    if len(s) != len(goal) {
        return false
    }

    sets := make(map[byte]int, 0)
    diffs := make([]int, 0)
    for i := 0; i < len(s); i++ {
        if s[i] != goal[i] {
            diffs = append(diffs, i)
        }
        sets[s[i]]++
    }

    if len(diffs) == 0 {
        for _, v := range sets {
            if v >= 2 {
                return true
            }
        }
    }

    return len(diffs) == 2 && s[diffs[0]] == goal[diffs[1]] && s[diffs[1]] == goal[diffs[0]]
}

File: 0889-buddy-strings/0889-buddy-strings.ts
function buddyStrings(s: string, goal: string): boolean {
    if (s.length !== goal.length) return false

    const m = {}

    for (let i = 0; i < s.length; i++) {
        const c = s[i]

        if (!m[c]) {
            m[c] = 1
        } else {
            m[c]++
        }
    }

    let canSwap = false
    let diffCnt = 0

    for (let i = 0; i < goal.length; i++) {
        if (goal[i] !== s[i]) {
            diffCnt++
            if (diffCnt > 2) return false
        }

        const c = goal[i]   
        const characterCnt = m[c]
        if(!characterCnt) return false

        // 'cc' => 'cc
        if (characterCnt > 1) {
            canSwap = true
        // 'cb' => 'bc'
        } else if (characterCnt === 1 && s[i] !== c) {
            canSwap = true
        }

        m[c]--
        if (m[c] < 0) return false
    }


    return canSwap
};